sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/export/MemoryMode.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.export;
+
+import io.opentelemetry.sdk.metrics.data.MetricData;
+import io.opentelemetry.sdk.metrics.internal.export.MetricProducer;
+
+/** The type of memory allocation used during metric collection in {@link MetricReader}. */
+public enum MemoryMode {
```

Comment:
That's a good question.. Theoretically it could be applicable to logs and spans as well.. 

If we did introduce a similar concept in other signals, I wonder if the memory modes would always overlap, or if we would want signal specific settings. If we made `MemoryMode` metric specific for now, could always introduce it to the other signals later with a copy.. Or we could take a bit of a gamble and put it in `opentelemetry-sdk-common`, anticipating that it would be useful in other signals and the enum values would be the same.

Reply:
I think the main question here is - what is repeatable in logs/traces?
In metrics - it's the data-point for each attributes. Attributes usually repeated when reported, hence it's easy.  In Async we're pooling the data points. Since I don't know at all the logs/traces, I'm wondering out loud, what repeats there, what are pooling?

Reply:
In traces and logs we could reuse the `SdkSpan` passed to `SpanProcessor`, and `SpanData` passed to `SpanExporter`. Similar story in logs. The main difference is that in metrics, most of the time applications should reach a steady state where the set streams levels out and the same set of `MetricData` is reused continuously for each collection. With spans and logs, we may reach a steady state where roughly the same number of spans / logs are produced for each collection, but it will be less predictable than metrics, so harder to tune the size of the object pools.

Reply:
> In traces and logs we could reuse the `SdkSpan` passed to `SpanProcessor`, and `SpanData` passed to `SpanExporter`. Similar story in logs.

Yeah, that's definitely the use case I had in mind.
(It might perhaps be possible to use similar technique in the OTLP marshallers themselves, but that's specific to the exporters, and would not influence `MemoryMode` in any way)

> The main difference is that in metrics, most of the time applications should reach a steady state where the set streams levels out and the same set of `MetricData` is reused continuously for each collection. With spans and logs, we may reach a steady state where roughly the same number of spans / logs are produced for each collection, but it will be less predictable than metrics, so harder to tune the size of the object pools.

That's true; still, I expect that pooling even some of the `SpanData`/`LogRecordData` objects might be very beneficial for applications that generate lots of telemetry.

Reply:
Yes. This suggests that we should move memory mode to `:sdk:common`, in the `io.opentelemetry.sdk.common.export` package: https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk/common/src/main/java/io/opentelemetry/sdk/common/export

Reply:
@jack-berg Ok, I have moved `MemoryMode` to :sdk:common in the export package.
Shall I move `ObjectPool` and `PooledHashMap` to :sdk:common as well? under which package?